stlab/future.hpp

```diff
@@ -1042,24 +1047,33 @@ namespace detail
             , _holds(_remaining)
         {}
 
-        std::atomic_size_t                    _remaining{0};
+        std::atomic_size_t                    _remaining;
         std::vector<future<void>>             _holds;
         std::mutex                            _mutex;
         boost::optional<std::exception_ptr>   _error;
-        size_t                                _index{0};
+        size_t                                _index{std::numeric_limits<size_t>::max()};
         packaged_task<>                       _f;
 
         void failure(std::exception_ptr error) {
-            // we already have a result
-            if (_remaining == 0) {
-                return;
-            }
+            size_t current_remaining = _remaining.load();
+            
+            bool run = false;
+            
             // only the last error is of any interest
-            if (_remaining == 1) {
-                _error = std::move(error);
-                _f();
+            while (current_remaining != 0) { // we already have a result
+                size_t new_remaining = current_remaining - 1;
+                if (new_remaining == 0) {    // this was the last chance
+                    _error = std::move(error); 
+                    run = true;
+                    break;
+                }
+
+                if (_remaining.compare_exchange_strong(current_remaining, new_remaining)) {
```

Comment:
If the remaining value is decremented between the load and compare_exchange_strong, then the decrement is ignored? That doesn't seem right. Anytime I see multiple atomic ops like this I get very nervous about correctness. The usual pattern for a count is to decrement and test i.e. if (--_remaining) ...

This whole loop business seems wrong. If current_remaining is some value other than one this is a spin lock? This library needs to run without threads... no blocking allowed.